### PR TITLE
Compute Quadrature formula over a general poly

### PR DESCRIPTION
**Pull request by fdrmrc**
Given two deal.II cells, these two functions allow to compute a Quadrature rule over the region described by a given boolean operation between these two. Internally, the polyhedron is split into simplices and `compute_affine_transformation()` is exploited in order to get a rule on each one of them. 

These will be used inside the `NonMatching` namespace to provide exact quadratures during the assembly of coupling matrices (like the one in step-60). In particular, an exact version of `NonMatching::create_coupling_mass_matrix` (https://www.dealii.org/developer/doxygen/deal.II/namespaceNonMatching.html#abe338fc879922cc67627b5ca039fc663) will be available.

Currently, this works in 3D and codimension 0, but I have working versions for 2D-2D, 1D-2D and 2D-3D.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]
This looks really nice. I must admit that I do not exactly understand the underlying algorithms in CGAL, but the overall workflow seems nice and convincing, so I agree with this one here. (It was a good opportunity for me to learn about the underlying assumptions in the CGAL part, like C++17 and similar.)

File: include/deal.II/cgal/utilities.h [hunk 1]
```diff
@@ -19,6 +19,12 @@
 #include <deal.II/base/config.h>
 
 #ifdef DEAL_II_WITH_CGAL
+#  include <deal.II/base/quadrature_lib.h>
+
+#  include <deal.II/grid/tria.h>
+
+#  include <boost/hana.hpp>
```
Comment (drwells) [inline include/deal.II/cgal/utilities.h:26]: We will need to record this as another external boost dependency - we don't rely on hana anywhere else.

File: include/deal.II/cgal/utilities.h [hunk 2]
```diff
@@ -198,6 +207,73 @@ namespace CGALWrappers
...
+   * a Quadrature rule to integrate over the polygon by looping trough all the
+   * vertices and exploiting QGaussSimplex.
+   *
+   * @param[in] tria The CGAL triangulation object describing the polyhedral
+   * region.
+   * @param[in] degree Desired degree of the Quadrature rule on each element of
+   * the polyhedral.
+   * @return A global Quadrature rule that allows to integrate over the polyhedron.
+   */
+  template <typename Tr>
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply include/deal.II/cgal/utilities.h:222]: Sure, thanks!

File: include/deal.II/cgal/utilities.h [hunk 3]
```diff
@@ -198,6 +207,73 @@ namespace CGALWrappers
...
+    const Mapping<dim0, spacedim> &mapping0 =
+      (dealii::ReferenceCells::get_hypercube<dim0>()
+         .template get_default_linear_mapping<dim0, spacedim>()),
+    const Mapping<dim1, spacedim> &mapping1 =
+      (dealii::ReferenceCells::get_hypercube<dim1>()
+         .template get_default_linear_mapping<dim1, spacedim>()));
+  /**
+   * A specialization of the function above when the BooleanOperation is an
+   * intersection. The rationale behind this specialization is that deal.II
+   * cells are convex sets, and as the intersection of convex sets is itself
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply include/deal.II/cgal/utilities.h]: Sorry for being late here. I've just document what really the code is doing, thanks for pointing that out.

**Comment by fdrmrc** [comment]
Since this is not ready yet for all possible dim-spacedim dimensions, I think we can postpone this to 9.5, together with #13814

**Comment by kronbichler** [comment]
The test failures are unrelated.

**Pull request merged**